### PR TITLE
fw_info structure change

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Bundle a few changes which modify the fw_info structure. These should ideally be bundled so the number of competing formats are as few as possible.

- Allow explicit invalidation of fw_info structs.
- Refactor EXT_APIs to provide more transparent versioning allowing external checking of version requirements. In the process, the fw_info.h file was reordered quite extensively.
- Add more supported offsets for fw_info structs (0x0 and 0x1000).

Additional changes:
- Add tests for fw_info (not complete coverage)

### CODE REVIEW

File: include/fw_info.h [hunk 1]
```diff
@@ -212,12 +226,25 @@ static inline const struct fw_info *fw_info_check(u32_t fw_info_addr)
...
-#define FW_INFO_OFFSET0 0x200
-#define FW_INFO_OFFSET1 0x400
-#define FW_INFO_OFFSET2 0x800
-#define FW_INFO_OFFSET_COUNT 3
+#define FW_INFO_OFFSET0 0x0
+#define FW_INFO_OFFSET1 0x200
+#define FW_INFO_OFFSET2 0x400
+#define FW_INFO_OFFSET3 0x800
+#define FW_INFO_OFFSET4 0x1000
+#define FW_INFO_OFFSET_COUNT 5
```
Comment (sigvartmh) [inline include/fw_info.h]: why do we now have 5?

Comment (oyvindronningstad) [reply include/fw_info.h]: It's one of the changes (Add more supported offsets).

Comment (sigvartmh) [reply include/fw_info.h]: okei thanks for clearing it up :)

File: include/fw_info.rst [hunk 2]
```diff
@@ -13,10 +13,10 @@ Purpose
 The purpose of the firmware information is to allow other images such as bootloaders, or infrastructure such as firmware servers, to gain information about the firmware image.
 
 The firmware information structure has a 12-byte magic header and a verified binary layout to ensure that the format is portable and identifiable.
-It must be located at one of three offsets from the start of the image: 0x200, 0x400, or 0x800.
+It must be located at one of three offsets from the start of the image: 0x0, 0x200, 0x400, 0x800, 0x1000.
```
Comment (ru-fu) [inline include/fw_info.rst]: one of the following

File: include/fw_info.rst [hunk 3]
```diff
@@ -29,6 +29,9 @@ It includes the following information:
 * The single, monotonically increasing version number of the image.
 * The address through which to boot into the firmware (the vector table address).
   This address is not necessarily the start of the image.
+* A value that must be 0xFFFFFFFF.
+  If this is set to 0 (e.g. by the bootloader), the image can quickly be established as invalid.
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply include/fw_info.rst]: In practice it needs only be different from `0xFFFFFFFF`.

File: include/fw_info.rst [hunk 4]
```diff
@@ -29,6 +29,9 @@ It includes the following information:
 * The single, monotonically increasing version number of the image.
 * The address through which to boot into the firmware (the vector table address).
   This address is not necessarily the start of the image.
+* A value that must be 0xFFFFFFFF.
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply include/fw_info.rst]: Can you elaborate? If this check is spoofed during validation, the signature check will still fail.

File: subsys/bootloader/bl_validation/bl_validation.c [hunk 5]
```diff
@@ -161,7 +161,7 @@ static bool validate_firmware(u32_t fw_dst_address, u32_t fw_src_address,
 					(u8_t *)fw_val_info->firmware_address,
 					fwinfo->firmware_size);
 
-		if (retval != -ESIGINV) {
+		if (retval != -EHASHINV) {
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/bootloader/bl_validation/bl_validation.c]: Yep, you're right :)

File: subsys/fw_info/Kconfig [hunk 6]
```diff
@@ -83,7 +83,7 @@ config FW_INFO_HARDWARE_ID
 config FW_INFO_VERSION
 	int "Version number of the fw_*_info structs (8 bits)."
 	range 0 255
-	default 1
+	default 2
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/fw_info/Kconfig]: will squash

File: include/fw_info.rst (hunk 2 above)
Comment (oyvindronningstad) [reply include/fw_info.rst]: Thanks, fixed.

**Comment by oyvindronningstad** [comment]
Created 3 other PRs:

#1386, #1387, #1388

File: include/fw_info.rst [hunk 7]
```diff
@@ -29,6 +29,9 @@ It includes the following information:
...
+  If this is set to 0 (e.g. by the bootloader), the image can quickly be established as invalid.
+  The bootloader will set this to 0 when the image fails validation in some way, so a costly validation doesn't need to be performed on every boot.
```
Comment (ru-fu) [inline include/fw_info.rst]:
```suggestion
  The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.
```

**Comment by hakonfam** [comment]
Typo in one of the commits: "Exlicitly"

File: include/fw_info.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (hakonfam) [reply include/fw_info.rst]: Since 0xFFFFFFFF is the  value of erased memory, my gut feeling is that it would be easier to spoof this value somehow, compared to a random value 0x2B68C12A

File: include/fw_info.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (hakonfam) [reply include/fw_info.rst]: If the code enforce that non `0xFFFFFFFF` implies invalid, then IMO we should state that in the docs, otherwise it could be considered a bug that `0xFFFFFFF0` is considered invalid.

File: subsys/bootloader/bl_validation/bl_validation.c [hunk 8]
```diff
@@ -113,6 +113,12 @@ static bool validate_firmware(u32_t fw_dst_address, u32_t fw_src_address,
 		return false;
 	}
 
+	if (fwinfo->valid != 0xFFFFFFFF) {
```
Comment (hakonfam) [inline subsys/bootloader/bl_validation/bl_validation.c]: Lets use some define instead of a magic value for `0xFFFFFFFF`

File: subsys/fw_info/Kconfig [hunk 9]
```diff
@@ -82,7 +82,7 @@ config FW_INFO_HARDWARE_ID
...
```
Comment (hakonfam) [inline subsys/fw_info/Kconfig]: On a side note, should this field even be visible for the users? IE should it have a prompt? IMO this could be considered a hidden option.

File: include/fw_info.h [hunk 10]
```diff
@@ -64,10 +72,10 @@ struct __packed fw_info {
 	u32_t firmware_address;
 
 	/* Where to place the getter for the ABI provided to this firmware. */
-	fw_info_abi_getter *abi_in;
+	struct fw_info_abi_getter_info *abi_in;
 
 	/* This firmware's ABI getter. */
-	const fw_info_abi_getter abi_out;
+	fw_info_abi_getter abi_out;
```
Comment (hakonfam) [inline include/fw_info.h]: Why is this no longer `const` ?

Comment (oyvindronningstad) [reply include/fw_info.h]: Because none of the other struct members are const. All relevant struct instantiations are marked const anyways, so this isn't strictly needed.

File: include/fw_info.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply include/fw_info.rst]: Updated docs

File: include/fw_info.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply include/fw_info.rst]: Ok, I think I understand. For it to work in the OTP region, it would have to be something like 

    #define VALID_VAL 0x2B68FFFF
    #define INVALID_VAL 0x2B680000

File: subsys/bootloader/bl_validation/bl_validation.c (hunk 8 above)
Comment (oyvindronningstad) [reply subsys/bootloader/bl_validation/bl_validation.c]: Done

File: subsys/fw_info/Kconfig (hunk 9 above)
Comment (oyvindronningstad) [reply subsys/fw_info/Kconfig]: Done

**Comment by oyvindronningstad** [comment]
> Typo in one of the commits: "Exlicitly"

Done

File: include/fw_info.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply include/fw_info.rst]: Implemented now. I also put the valid value in Kconfig.

[the remark this answers is not in the document]
Comment (ru-fu) [reply include/fw_info.rst]: Does this need a doc update?

[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply include/fw_info.rst]: Yes, thanks. Updated docs now.

**Comment by oyvindronningstad** [comment]
I refactored the ABI functionality. See the ABI commit.

**Comment by oyvindronningstad** [comment]
Also, I separated the fix-commit into a new branch. Will open a separate PR.

File: subsys/fw_info/fw_info.c [hunk 11]
```diff
@@ -80,3 +92,15 @@ const struct fw_info_abi *fw_info_abi_find(u32_t id, u32_t flags,
 	return NULL;
 }
 
+
+#define INVALID_VAL 0xFFFF0000
+
+#ifdef CONFIG_NRFX_NVMC
```
[the remark this answers is not in the document]
Comment (hakonfam) [reply subsys/fw_info/fw_info.c]: Note this comment might be outdated, I see that I forgot to submit it, and it was done before the re-wriet.

[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: You can want to have FW_INFO without the invalidation logic, which is why I didn't make it `depends on`. Note that FW_INFO has `imply NRFX_NVMC`.

**Comment by hakonfam** [comment]
https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/238 needs to be merged just before this gets merged.

File: include/fw_info.h [hunk 12]
```diff
@@ -27,22 +27,36 @@ extern "C" {
...
- *                       retrieve the different ones with this.
- * @param[out]   abi     Pointer to the abi with the given id and index.
- *
- * @retval 0        Success.
- * @retval -ENOENT  id not found.
- * @retval -EBADF   index too large.
- * @retval -EFAULT  abi was NULL.
+
+/**
+ * This struct is used to request an ABI. The bootloader will populate the
```
Comment (hakonfam) [inline include/fw_info.h]: Isn't it possible for any image to provide an ABI? Here it states `The bootloader will populate`

Comment (oyvindronningstad) [reply include/fw_info.h]: The "bootloader" here is any image that boots another image :).
I could change it to "A bootloader will...". Suggestions?

File: include/fw_info.h [hunk 13]
```diff
@@ -79,12 +103,22 @@ struct __packed fw_info {
...
+OFFSET_CHECK(struct fw_info, firmware_version, 16);
+OFFSET_CHECK(struct fw_info, firmware_address, 20);
+OFFSET_CHECK(struct fw_info, valid, 24);
+OFFSET_CHECK(struct fw_info, abi_out_len, 28);
+OFFSET_CHECK(struct fw_info, abi_out, 32);
+OFFSET_CHECK(struct fw_info, abi_in_len, 36);
+OFFSET_CHECK(struct fw_info, abi_in, 40);
+OFFSET_CHECK(struct fw_info, reserved00, 44);
+
+OFFSET_CHECK(struct fw_info_abi_request, abi_id, 0);
```
Comment (frkv) [inline include/fw_info.h]: Could these static asserts be closer to the type declaration? It comes out of order with the other type declaration first fw_info and then fw_info_abi_request

File: include/fw_info.h [hunk 14]
```diff
@@ -98,8 +132,10 @@ OFFSET_CHECK(struct fw_info, firmware_address,
  * ID/version combination.
```
Comment (frkv) [inline include/fw_info.h]: Comment on a description and  name of this type:
It seems we spend most of the documentation on fw_info_abi talking about what this type does *not* include in long  related sentences. I would name this a "header" in the type name or use some other naming to emphasize that this is not the complete structure for the ABI. 

Alternatively the documentation could state something like this
This struct serves as the header for an ABI, containing only the basic information about the ABI without describing the  implementation details dependent on an ID and version combination.

File: include/fw_info.h [hunk 15]
```diff
@@ -311,6 +346,16 @@ const struct fw_info_abi *fw_info_abi_get(u32_t id, u32_t index);
 const struct fw_info_abi *fw_info_abi_find(u32_t id, u32_t flags,
 					u32_t min_version, u32_t max_version);
 
+
+/**Invalidate this image by setting the @c valid value to 0x0.
+ *
+ * @note This function needs to have CONFIG_NRF_NVMC enabled.
+ *
+ * @param[in]  fw_info  The info structure to modify.
```
Comment (frkv) [inline include/fw_info.h]: Modify is the mechanism, maybe use "invalidate" to be clear about the what the act of modifying this structure specifically is doing?

File: include/fw_info.h [hunk 16]
```diff
@@ -214,41 +277,48 @@ static inline const struct fw_info *fw_info_check(u32_t fw_info_addr)
...
 
-#define CURRENT_OFFSET (CONFIG_FW_INFO_OFFSET + VECTOR_OFFSET)
+#define FW_INFO_CURRENT_OFFSET (CONFIG_FW_INFO_OFFSET + FW_INFO_VECTOR_OFFSET)
 
-static const u32_t allowed_offsets[] = {FW_INFO_OFFSET0, FW_INFO_OFFSET1,
-					FW_INFO_OFFSET2};
+static const u32_t fw_info_allowed_offsets[] = {
+					FW_INFO_OFFSET0, FW_INFO_OFFSET1,
+					FW_INFO_OFFSET2, FW_INFO_OFFSET3,
+					FW_INFO_OFFSET4};
```
Comment (frkv) [inline include/fw_info.h]: If the FW_INFO_OFFSET is always increasing array, what about making use of a FW_INFO_OFFSET_INVALID_LAST at the end? This would simplify the logic for the count further down.

File: include/fw_info.h [hunk 17]
```diff
@@ -214,41 +277,48 @@ static inline const struct fw_info *fw_info_check(u32_t fw_info_addr)
...
-BUILD_ASSERT_MSG(ARRAY_SIZE(allowed_offsets) == FW_INFO_OFFSET_COUNT,
+BUILD_ASSERT_MSG(ARRAY_SIZE(fw_info_allowed_offsets) == FW_INFO_OFFSET_COUNT,
 		"Mismatch in the number of allowed offsets.");
 /** @endcond
  */
 
-#if (FW_INFO_OFFSET_COUNT != 3) || ((CURRENT_OFFSET) != (FW_INFO_OFFSET0) && \
-				(CURRENT_OFFSET) != (FW_INFO_OFFSET1) && \
-				(CURRENT_OFFSET) != (FW_INFO_OFFSET2))
+#if (FW_INFO_OFFSET_COUNT != 5) \
```
Comment (frkv) [inline include/fw_info.h]: See comment on FW_INFO_OFFSET_INVALID_LAST

File: include/fw_info.h [hunk 18]
```diff
@@ -259,7 +329,7 @@ static inline const struct fw_info *fw_info_find(u32_t firmware_address)
 
 	for (u32_t i = 0; i < FW_INFO_OFFSET_COUNT; i++) {
```
Comment (frkv) [inline include/fw_info.h]: See comment on FW_INFO_OFFSET_INVALID_LAST

File: include/fw_info.rst [hunk 19]
```diff
@@ -29,6 +29,9 @@ It includes the following information:
 * The single, monotonically increasing version number of the image.
 * The address through which to boot into the firmware (the vector table address).
   This address is not necessarily the start of the image.
+* A value that can be modified in place to invalidate the firmware. See :option:`CONFIG_FW_INFO_VALID_VAL`.
+  If this is set to any other value than :option:`CONFIG_FW_INFO_VALID_VAL` (e.g. by the bootloader), the image can quickly be established as invalid.
+  The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.
```
Comment (frkv) [inline include/fw_info.rst]: Maybe add a point on this value being guarded from denial-of-service invalidation by a third party

File: samples/bootloader/prj.conf [hunk 20]
```diff
@@ -16,6 +17,10 @@ CONFIG_NRF_RTC_TIMER=n
 CONFIG_CLOCK_CONTROL=n
 CONFIG_FPROTECT=y
 CONFIG_SECURE_BOOT_DEBUG=y
+CONFIG_ASSERT=y
+CONFIG_BL_ROT_VERIFY_ABI_PROVIDE=y
```
Comment (frkv) [inline samples/bootloader/prj.conf]: I have not seen the verb "provide" used in this context. Is this Zephyr nomenclature?

What about _PROVIDED, _ENABLED, _SUPPORT, or _SUPPORTED?

File: include/fw_info.h [hunk 21]
```diff
@@ -78,75 +60,156 @@ struct __packed fw_info {
...
+/** @endcond
+ */
+
+
+/**
+ * This struct is used to request an ABI. It contains a pointer to a non-
+ * initialized pointer that a previous boot stage will populate with a pointer
+ * to a compliant ABI in another image. An ABI fulfills a request if the ID
+ * matches, all flags in the request are set in the ABI, and the version falls
+ * between the minimum and maximum (inclusive). If `required` is true the image
```
Comment (frkv) [inline include/fw_info.h]: Comma needed: If required is true (comma) the image will not function

File: include/fw_info.h [hunk 22]
```diff
@@ -311,6 +375,16 @@ const struct fw_info_abi *fw_info_abi_get(u32_t id, u32_t index);
 const struct fw_info_abi *fw_info_abi_find(u32_t id, u32_t flags,
 					u32_t min_version, u32_t max_version);
 
+
+/**Invalidate this image by setting the @c valid value to 0x0.
```
Comment (frkv) [inline include/fw_info.h]: Possible difficulty of matching "image" to content inside fw_info (the valid flag)
since the param[in] fw_info is talking about invalidating the info structure

Suggestion:

Function that invalidates an image by manipulating fw_info

@details Setting the @c valid value to 0x0 in fw_info, invalidates the image it represents

File: include/fw_info.h [hunk 23]
```diff
@@ -78,75 +60,156 @@ struct __packed fw_info {
...
+OFFSET_CHECK(struct fw_info_abi, abi_len, 0);
+OFFSET_CHECK(struct fw_info_abi, abi_id, 4);
+OFFSET_CHECK(struct fw_info_abi, abi_flags, 8);
+OFFSET_CHECK(struct fw_info_abi, abi_version, 12);
+/** @endcond
+ */
+
+
+/**
+ * This struct is used to request an ABI. It contains a pointer to a non-
```
Comment (frkv) [inline include/fw_info.h]: I would suggest a @details field instead of all the details in one long text

/**
 * Structure used to request an ABI
*
* @details ....
*
* @note If `required` is true the image will not function if it has no access to the ABI.
*/

I would also recommend moving the information about the pointer to a non-initialized pointer to the structure member **abi** instead, to improve readibility. If this is done, the ingress could talk about generic functionality of this struct

File: include/fw_info.h [hunk 24]
```diff
@@ -78,75 +60,156 @@ struct __packed fw_info {
...
+ * initialized pointer that a previous boot stage will populate with a pointer
+ * to a compliant ABI in another image. An ABI fulfills a request if the ID
```
Comment (frkv) [inline include/fw_info.h]: Not easy to follow the difference and connection between the "previous boot stage", "another image" and "the image" in this explaination

File: include/fw_info.h [hunk 25]
```diff
@@ -27,47 +27,29 @@ extern "C" {
...
 /**
- * This is a data structure that is placed at a specific offset inside a
- * firmware image so it can be consistently read by external parties. The
- * specific offset makes it easy to find, and the magic value at the start
- * guarantees that it contains data of a specific format.
+ * This struct serves as a header before an actual ABI. This header contains
+ * basic metadata about the ABI: A unique identifier, version and flags to be
+ * used for compatibility matching. The payload that comes after this header is
+ * completely implementation dependent for each ABI.
+ * Note: How to use the ABI, such as the signatures of all the functions in the
```
Comment (frkv) [inline include/fw_info.h]: Express as actual @note

File: samples/bootloader/src/main.c [hunk 26]
```diff
@@ -46,14 +46,15 @@ extern u32_t _vector_table_pointer;
...
 {
-	u32_t *vector_table = (u32_t *)fw_info->firmware_address;
+	u32_t *vector_table = (u32_t *)fw_info->image_address;
 
 	printk("Attempting to boot from address 0x%x.\n\r",
-		fw_info->firmware_address);
+		fw_info->image_address);
 
-	if (!bl_validate_firmware_local(fw_info->firmware_address,
+	if (!bl_validate_firmware_local(fw_info->image_address,
```
Comment (frkv) [inline samples/bootloader/src/main.c]: bl_validate_firmware_local is not documented (but not part of this PR)

I would suggest indicating that fw_info_invalidate is required to be run when this is false (as this function is also used by prevalidate so FW can't be invalidated directly)

File: tests/subsys/fw_info/src/main.c [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+	.image_version = 0xFF,
+	.valid = CONFIG_FW_INFO_VALID_VAL,
+};
+
+u8_t fw_info_find_buf[0x1000 + sizeof(dummy_s1)];
+
+u32_t dummy_align;
+u8_t memeq_buf1[] = {1, 2, 3, 4, 5, 6, 7, 8}; /* Aligned */
+u8_t memeq_buf2[] = {1, 2, 3, 4, 5, 6, 7, 8}; /* Aligned */
+u8_t dummy_misalign;
```
Comment (frkv) [inline tests/subsys/fw_info/src/main.c]: I don't know if this leads to an unaligned type behind it. Maybe try with byte-string buffers that are unaligned

File: samples/bootloader/prj.conf (hunk 20 above)
Comment (oyvindronningstad) [reply samples/bootloader/prj.conf]: How about instead of _PROVIDE and _USE we have _ENABLED and _REQUIRED?

File: include/fw_info.rst (hunk 19 above)
Comment (oyvindronningstad) [reply include/fw_info.rst]: Added a sentence

File: samples/bootloader/prj.conf (hunk 20 above)
Comment (oyvindronningstad) [reply samples/bootloader/prj.conf]: Implemented this now. Take a look to see if it works.

File: include/fw_info.h (hunk 22 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Changed to

    /**Invalidate an image by manipulating its fw_info.
     * 
     * @details Invalidation happens by setting the @c valid value to 0x0.

File: include/fw_info.h (hunk 23 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Added `@details`.

File: include/fw_info.h (hunk 24 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Rewrote this section

File: samples/bootloader/src/main.c (hunk 26 above)
Comment (oyvindronningstad) [reply samples/bootloader/src/main.c]: Added dox for `bl_validate_firmware_local()`

File: tests/subsys/fw_info/src/main.c (hunk 27 above)
Comment (oyvindronningstad) [reply tests/subsys/fw_info/src/main.c]: You were right, thanks! Fixed this now and added asserts for the expected alignment.

File: include/fw_info.h (hunk 15 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Done

File: include/fw_info.h (hunk 14 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: This has been rewritten, take another look if you haven't already.

File: include/fw_info.h (hunk 13 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Moved stuff around.

File: include/fw_info.h (hunk 12 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: This is now described as "a previous boot stage".

File: samples/bootloader/src/main.c (hunk 26 above)
Comment (frkv) [reply samples/bootloader/src/main.c]: Looks good

File: samples/bootloader/prj.conf (hunk 20 above)
Comment (frkv) [reply samples/bootloader/prj.conf]: Yes. Easier to read, for sure... Accepted

File: include/fw_info.rst [hunk 28]
```diff
@@ -29,91 +29,142 @@ It includes the following information:
...
+  If the firmware is write protected before being booted by the bootloader, only the bootloader will be able to invalidate it.
```
Comment (ru-fu) [inline include/fw_info.rst]:
```suggestion
  If the firmware is write-protected before being booted by the bootloader, only the bootloader can invalidate it.
```

**Comment by oyvindronningstad** [comment]
Placed some changes into separate PRs: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1678, https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1679

**Comment by oyvindronningstad** [comment]
DNM until #1685  is merged.

**Comment by oyvindronningstad** [comment]
@ru-fu There's been a few more changes. Here is the diff of fw_info.rst since your last comment (Dec 13):

https://github.com/NordicPlayground/fw-nrfconnect-nrf/compare/d1d5e588ac99bf6372212a3a5aba48fe07127485..c1ded6042cf0038b52f290cc32cd76967168373f#diff-a911f1d849bdca671f0c2e653412ce5a

It's mostly an additional paragraph on binary compatibility.

File: include/fw_info.h [hunk 29]
```diff
@@ -274,6 +280,17 @@ const struct fw_info_ext_api *fw_info_ext_api_get(u32_t id, u32_t index);
 const struct fw_info_ext_api *fw_info_ext_api_find(u32_t id, u32_t flags,
 					u32_t min_version, u32_t max_version);
 
+/**Invalidate an image by manipulating its fw_info.
+ *
+ * @details Invalidation happens by setting the @c valid value to 0x0.
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply include/fw_info.h]: It makes "valid" monospace. http://www.doxygen.nl/manual/commands.html#cmdc

File: subsys/fw_info/fw_info.c [hunk 30]
```diff
@@ -81,3 +83,15 @@ const struct fw_info_ext_api *fw_info_ext_api_find(u32_t id, u32_t flags,
...
+void fw_info_invalidate(const struct fw_info *fw_info)
+{
+	/* Check if value has been written. */
+	if (fw_info->valid == CONFIG_FW_INFO_VALID_VAL) {
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Discussed offline. Will leave as is, since this check and the check for valid images are consistent.

File: subsys/fw_info/Kconfig [hunk 31]
```diff
@@ -101,4 +102,10 @@ config FW_INFO_MAGIC_LEN
 	  Total length of magic words (in bytes).
 	  For more information, see FW_INFO_MAGIC_COMMON.
 
+config FW_INFO_VALID_VAL
+	hex
+	default 0x9102FFFF
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/fw_info/Kconfig]: Thanks, fixed.

File: subsys/fw_info/fw_info.c [hunk 32]
```diff
@@ -81,3 +83,15 @@ const struct fw_info_ext_api *fw_info_ext_api_find(u32_t id, u32_t flags,
 	return NULL;
 }
 
+
+#define INVALID_VAL 0xFFFF0000
```
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Done :)

File: include/fw_info.h [hunk 33]
```diff
@@ -87,81 +69,199 @@ struct __packed fw_info {
...
+ * @note If `required` is true, the image making the request will not function
+ *       unless it has access to the EXT_API.
+ *
+ */
+struct __packed fw_info_ext_api_request {
+	/* The minimum EXT_API version. */
+	struct fw_info_ext_api min_ext_api;
+
+	/* The maximum version accepted. */
+	u32_t ext_api_max_version;
```
Comment (hakonfam) [inline include/fw_info.h]: Could you explain why the min version is a struct while the max version is an int?

Comment (oyvindronningstad) [reply include/fw_info.h]: `min_ext_api` refers to the minimum implementation, wrt both version and flags.

File: include/fw_info.h [hunk 34]
```diff
@@ -248,51 +348,19 @@ static inline const struct fw_info *fw_info_find(u32_t firmware_address)
...
- *                          version.
+ * @param[in]  fw_info    Pointer to the other firmware's information structure.
+ * @param[in]  test_only  Don't populate ext_api_in.
  *
- * @return The EXT_API, or NULL if none was found.
+ * @return Whether requirements could be satisified.
  */
-const struct fw_info_ext_api *fw_info_ext_api_find(u32_t id, u32_t flags,
-					u32_t min_version, u32_t max_version);
+bool fw_info_ext_api_provide(const struct fw_info *fwinfo, bool test_only);
```
Comment (hakonfam) [inline include/fw_info.h]: I don't like having code for testing only. Could this be handled in a different way? e.g. macro, creating a wrapper or something?

File: include/fw_info.rst [hunk 35]
```diff
@@ -53,20 +54,35 @@ The header consists of the following information:
...
-This EXT_API getter must be provided when booting the image.
-In other words, an image should expect its :cpp:member:`ext_api_in` EXT_API getter to be filled at the time it boots, and will not touch it during booting.
-After booting, an image can call :cpp:member:`ext_api_in` to retrieve EXT_APIs from the image or images that booted it without knowing where they are located.
+An EXT_API request structure is a structure consiting of a description of the requested EXT_API in the form of an EXT_API header, plus a few more pieces of data.
+The structure also contains the location of a RAM buffer into which a pointer to a matching EXT_API can be placed before booting.
+The RAM buffer must not be initialized, or otherwise touched during bootup of the image.
+This is to preserve the value written there before boot.
+The EXT_API system allows a bootloader to parse the image's requests, the parse its (the bootloader's) own, or other images' EXT_APIs to find matches, then populate the RAM buffers with pointers to other images' EXT_APIs.
+The bootloader will in this way also be aware of whether all images have access to their dependencies.
```
Comment (hakonfam) [inline include/fw_info.rst]: Check this sentence.

File: subsys/fw_info/fw_info.c [hunk 36]
```diff
@@ -6,84 +6,150 @@
...
+#include <sys/util.h>
+#include <init.h>
 #include <errno.h>
 #include <string.h>
 #include <nrfx_nvmc.h>
 
 
-extern const u32_t _image_rom_start;
-extern const u32_t _flash_used;
+extern const u32_t _image_rom_start[];
```
Comment (hakonfam) [inline subsys/fw_info/fw_info.c]: Maybe we should add a comment here stating that these are from the linkerscript? I guess not all readers will be known with this scheme.

File: include/fw_info.h (hunk 34 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: It refers to testing for the existence of the ext_apis. To be clear, the argument is meant to be used in regular code, not in tests. We can probably give it a better name.

Comment (hakonfam) [reply include/fw_info.h]: I only assumed that it was only used for testing.

To me "test only" heavily implies that this is only used for testing, but perhaps that is just me :) Up to you if you want to rename it.

**Comment by oyvindronningstad** [comment]
> I need to see some tests using the EXT_API feature before approving :)

The existing bl_crypto tests call the crypto functions using EXT_APIs

File: include/fw_info.h (hunk 33 above)
Comment (hakonfam) [reply include/fw_info.h]: so the "min" doesn't really count for flags, right? IMHO the "min" here can be confusing for the reader, when this really is a set of requirements, not all of which has to be "bigger", if you get what I mean.

Comment (oyvindronningstad) [reply include/fw_info.h]: I agree, we should find a different name.
`requirements`?
`request`?
`ext_api` and then document how it should be interpreted?

File: include/fw_info.h (hunk 34 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: I'll try to find another name :)

File: include/fw_info.h (hunk 33 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Done (`ext_api`)

File: include/fw_info.h (hunk 34 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Done (`provide`)

File: subsys/fw_info/fw_info.c (hunk 36 above)
Comment (oyvindronningstad) [reply subsys/fw_info/fw_info.c]: Good idea, added.

File: include/fw_info.h (hunk 33 above)
Comment (oyvindronningstad) [reply include/fw_info.h]: Oops. Not done, need to fix some more.

Comment (oyvindronningstad) [reply include/fw_info.h]: Done (`request`)

File: include/fw_info.rst [hunk 37]
```diff
@@ -53,20 +54,35 @@ The header consists of the following information:
...
-This EXT_API getter is a function pointer that retrieves EXT_API structs (or rather pointers to EXT_API structs).
+An image's EXT_APIs are placed at the end of the information structure, immediately before its EXT_API requests.
 
-In addition, every image can access other EXT_APIs through a second EXT_API getter (:cpp:member:`ext_api_in`).
-This EXT_API getter must be provided when booting the image.
-In other words, an image should expect its :cpp:member:`ext_api_in` EXT_API getter to be filled at the time it boots, and will not touch it during booting.
-After booting, an image can call :cpp:member:`ext_api_in` to retrieve EXT_APIs from the image or images that booted it without knowing where they are located.
+An EXT_API request structure is a structure consiting of a description of the requested EXT_API in the form of an EXT_API header, plus a few more pieces of data.
+The structure also contains the location of a RAM buffer into which a pointer to a matching EXT_API can be placed before booting.
+The RAM buffer must not be initialized, or otherwise touched during bootup of the image.
```
Comment (frkv) [inline include/fw_info.rst]: Maybe add a helpful hint, like __attribute__((noinit))

Comment (oyvindronningstad) [reply include/fw_info.rst]: Added a sentence.

**Comment by hakonfam** [comment]
@oyvindronningstad  please rebase

**Comment by oyvindronningstad** [comment]
> @oyvindronningstad please rebase

Done

**Comment by oyvindronningstad** [comment]
@ru-fu FYI this is now only missing a final doc review.

**Comment by oyvindronningstad** [comment]
> Please check the commit and squash if everything looks okay.

Thanks @ru-fu! I squashed, and added a slight change: https://github.com/NordicPlayground/fw-nrfconnect-nrf/compare/a7704efbcc8b2b6ec5ba83660127efdba221854e..c012ec46c3f3b4e66eaec0f26d8a2a283aefea40

File: include/fw_info.rst [hunk 38]
```diff
@@ -82,38 +107,86 @@ Creating EXT_APIs
...
 To create an EXT_API, complete the following steps:
 
-1. Declare a new struct type that starts with the :c:type:`fw_info_ext_api` struct:
+1. Create a unique ID for the EXT_API:
+
+   .. code-block:: c
+
+      #define MY_EXT_API_ID 0xBEEF
+
+#. Create a Kconfig entries using :file:`Kconfig.template.fw_info_ext_api`:
```
Comment (ru-fu) [inline include/fw_info.rst]:
```suggestion
#. Create Kconfig entries using :file:`Kconfig.template.fw_info_ext_api`:
```

Comment (oyvindronningstad) [reply include/fw_info.rst]: Done :)

**Comment by hakonfam** [comment]
Please rebase, LETS DO THIS 💃

**Comment by oyvindronningstad** [comment]
> Please rebase, LETS DO THIS 💃

Still blocked by https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1750 :D

**Comment by tejlmand** [comment]
DNM until https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/238 is merged, and `west.yml` has been updated to correct SHA

**Comment by oyvindronningstad** [comment]
> DNM until [NordicPlayground/fw-nrfconnect-zephyr#238](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/238) is merged, and `west.yml` has been updated to correct SHA

The plan is to wait for #1750 at which point [NordicPlayground/fw-nrfconnect-zephyr#238] is no longer needed.

**Comment by oyvindronningstad** [comment]
Rebased now that #1750 is merged.

**Comment by lemrey** [comment]
Update the changelog if needed :)

**Comment by oyvindronningstad** [comment]
> Update the changelog if needed :)

Thanks :), done
